### PR TITLE
az iot central device purge messages. Purges cloud to device messages from queue

### PR DESCRIPTION
**Pull request by gtrifonov**
---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

image

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

File: azext_iot/central/services/device.py [hunk 1]
```diff
@@ -616,3 +616,42 @@ def run_manual_failback(
...
 
     return _utility.try_extract_result(response)
+
+
+def purge_c2d_messages(
+    cmd,
+    app_id: str,
+    device_id: str,
+    token: str,
+    api_version: str,
```
Comment (yang-hai-feng) [inline azext_iot/central/services/device.py:626]: Do we need api-version here?

File: azext_iot/tests/central/test_iot_central_devices_int.py [hunk 2]
```diff
@@ -97,7 +123,7 @@ def test_device_connect(self):
         command = "iot central device compute-device-key --pk {} -d {}".format(
             APP_PRIMARY_KEY, device_id
         )
-        device_primary_key = self.cmd(
+        device_primary_key = self.cmd_withoutToken(
```
Comment (lucadruda) [inline azext_iot/tests/central/test_iot_central_devices_int.py:126]: "compute-device-key" does not use api-version as well. consider to create a sort of "cmd_with_no_params" or add "token" as an ignored parameter to the command

**Review by lucadruda — DISMISSED** [review]

File: HISTORY.rst [hunk 3]
```diff
@@ -3,6 +3,14 @@
 Release History
 ===============
 
+0.12.2
++++++++++++++++
+
+**IoT Central updates**
+
+* Added `az iot central device purge messages ` to purge cloud to device message queue
```
Comment (digimaun) [inline HISTORY.rst:11]: The command action is defined as `purge-messages`, also (nit) extra space in between `.

File: azext_iot/central/_help.py [hunk 4]
```diff
@@ -397,6 +397,21 @@ def _load_central_devices_help():
             --device-id {deviceid}
     """
 
+    helps[
+        "iot central device purge-messages"
```
Comment (digimaun) [inline azext_iot/central/_help.py:401]: If the goal is to support more c2d commands in the future, consider an additional command group encompassing them to improve organization and reduce "verb-noun" commands. For example `az iot central device c2d-message purge`.

There is an existing c2d message command group in `az iot device`, though its focused for IoT Hub interaction which illustrates this idea. Pasting screenshot here for convenience:

 
image

File: azext_iot/central/params.py [hunk 5]
```diff
@@ -70,8 +70,8 @@ def load_central_arguments(self, _):
         context.argument(
             "device_id",
             options_list=["--device-id", "-d"],
-            help="The ID of the target device, "
-            "You can find the Device Id by clicking on the Connect button on the Device Details page.",
+            help="The device ID of the target device."
+            "You can find the device ID by, clicking on the Connect button on the Device Details page.",
```
Comment (digimaun) [inline azext_iot/central/params.py:74]: Ideally you don't need to switch experiences to determine this information. As in the CLI should be able to show/list device IDs.

File: azext_iot/central/services/device.py [hunk 6]
```diff
@@ -616,3 +616,36 @@ def run_manual_failback(
...
+        central_dns_suffix: {centralDnsSuffixInPath} as found in docs
+
+    Returns:
+        {
+            message: 'Cloud to device (C2D) message queue purged for device {device_id}.\\n
+            Total messages purged: {totalMessagesPurged}.'
+        } on success
+        Raises error on failure
+    """
+    url = "https://{}.{}/{}/{}/c2d".format(
```
Comment (digimaun) [inline azext_iot/central/services/device.py:646]: API version is not being used here, does the corresponding param need to be exposed to users?

Comment (gtrifonov) [reply azext_iot/central/services/device.py:646]: no, there is no versioning for following api's

Comment (digimaun) [reply azext_iot/central/services/device.py:646]: Ok. So presumably we can remove the API version at the command level to simplify the command surface.

image

File: azext_iot/central/command_map.py [hunk 7]
```diff
@@ -151,6 +151,12 @@ def load_central_commands(self, _):
             "get_device_twin",
         )
 
+    with self.command_group(
+        "iot central device c2d-message",
+        command_type=central_device_ops,
+    ) as cmd_group:
+        cmd_group.show_command("purge", "purge_c2d_messages")
```
Comment (digimaun) [inline azext_iot/central/command_map.py:158]: Should be  `cmd_group.command("purge", "purge_c2d_messages")` (not a show_command)

**Review by digimaun — APPROVED** [review]

**Pull request merged**